kits/daisy/src/robot/hexapod.cpp

```diff
@@ -0,0 +1,610 @@
+}
+
+std::unique_ptr<Hexapod> Hexapod::createDummy(const HexapodParameters& params)
+{
+  HexapodErrors hex_errors;
+  return std::unique_ptr<Hexapod>(
+    new Hexapod(std::shared_ptr<hebi::Group>(),
+                std::shared_ptr<hebi::Group>(),
+                std::shared_ptr<hebi::Group>(),
+                params, {}, hex_errors));
+}
+
+// TODO: think about making this non-const and returning limited values (and/or
+// having another function to do this) (or anything that uses this in the main
+// function retrieves it from the hexapod)
+void Hexapod::updateStance(const Eigen::Vector3d& translation_velocity, const Eigen::Vector3d& rotation_velocity, double dt)
+{
+  Eigen::Vector3d trans_vel_limited = translation_velocity;
+  // Cap Z velocity if we are too high
+  double cur_z = legs_[0]->getLevelHomeStanceZ();
+  double dz = trans_vel_limited(2) * dt;
+  if (cur_z + dz > params_.max_z_)
+    trans_vel_limited(2) = (params_.max_z_ - cur_z) / dt;
+  if (cur_z + dz < params_.min_z_)
+    trans_vel_limited(2) = (params_.min_z_ - cur_z) / dt;
+
+  Eigen::Vector3d rot_vel_limited = rotation_velocity;
+  // Don't tilt in step!
+  if (mode_ == Mode::Step)
+    rot_vel_limited(1) = 0;
+
+  Eigen::VectorXd all_angles(num_angles_);
+  {
+    std::lock_guard<std::mutex> guard(fbk_lock_);
+    all_angles = positions_;
+  }
+  int num_joints = Leg::getNumJoints();
+  Eigen::VectorXd current_leg_angles(num_joints);
+  for (int i = 0; i < num_legs_; ++i)
+  {
+    // subsample from full vector
+    current_leg_angles = all_angles.segment(num_joints * i, num_joints);
+    // Update stance of leg
+    legs_[i]->updateStance(trans_vel_limited, rot_vel_limited, current_leg_angles, dt);
+  }
+  vel_xyz_ = trans_vel_limited;
+}
+
+void Hexapod::setCommand(int leg_index, const VectorXd* angles, const VectorXd* vels, const VectorXd* torques)
+{
+  int num_joints = Leg::getNumJoints();
+
+  // For fake legs, just update positions vector:
+  if (real_legs_.count(leg_index) == 0)
+  {
+    if (angles != nullptr)
+    {
+      int leg_offset = leg_index * num_joints;
+      std::lock_guard<std::mutex> guard(fbk_lock_);
+      positions_.segment(leg_offset, num_joints) = *angles;
+    }
+    return;
+  }
+  // Get leg offset, taking into account all legs may not be present. First,
+  // count the number of legs before this one:
+  int legs_prev = std::count_if(real_legs_.begin(), real_legs_.end(),
+    [leg_index](int real_leg_idx) {return real_leg_idx < leg_index; });
+  int leg_offset = legs_prev * num_joints;
+  if (angles != nullptr)
+  {
+    assert(angles->size() == num_joints);
+    for (int i = 0; i < num_joints; ++i)
+      cmd_[leg_offset + i].actuator().position().set((*angles)[i]);
+  }
+  if (vels != nullptr)
+  {
+    assert(vels->size() == num_joints);
+    for (int i = 0; i < num_joints; ++i)
+      cmd_[leg_offset + i].actuator().velocity().set((*vels)[i]);
```

Comment:
See comment above, also see above for `effort()` below

Reply:
See explanation above.